File: 36.valid-sudoku.cpp
/*
 * @lc app=leetcode id=36 lang=cpp
 *
 * [36] Valid Sudoku
 */

// @lc code=start
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {

        for (vector<char> line : board) {
            int result = isValidBox(line);
            if (!result)
                return false;
        }

        for (int i = 0; i < 9; i++) {
            vector<char> line;
            for (int j = 0; j < 9; j++) {
                line.push_back(board[j][i]);
            }
            int result = isValidBox(line);
            if (!result)
                return false;
        }

        int x = 0, y = 0;
        for (int x = 0; x < 9; x += 3) {
            for (int y = 0; y < 9; y += 3) {
                vector<char> line;
                for (int i = x; i < x + 3; i++) {
                    for (int j = y; j < y + 3; j++) {
                        line.push_back(board[i][j]);
                    }
                }
                bool result = isValidBox(line);
                if (!result)
                    return false;
            }
        }

        return true;
    }

    bool isValidBox(vector<char> box) {
        set<char> mySet;
        for (char c : box) {
            if (c == '.')
                continue;
            else if (mySet.find(c) == mySet.end()) {
                mySet.insert(c);
            } else
                return false;
        }
        return true;
    }
};
// @lc code=end

File: 36.valid-sudoku.py
#
# @lc app=leetcode id=36 lang=python3
#
# [36] Valid Sudoku
#

# @lc code=start
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        arr = [[set() for _ in range(9)] for _ in range(3)]
        for i in range(9):
            for j in range(9):
                charr = board[i][j]
                if charr == '.': 
                    continue

                boxIndex = (i // 3)*3 + (j // 3)

                if charr in arr[0][i] or charr in arr[1][j] or charr in arr[2][boxIndex]:
                    return False

                arr[0][i].add(charr)
                arr[1][j].add(charr)
                arr[2][boxIndex].add(charr)
        
        return True

# @lc code=end

